datafusion/execution/src/cache/list_files_cache.rs

```diff
@@ -0,0 +1,731 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
```

Comment:
I thought the `cache_unit.rs` file was already getting pretty large so I pulled this implementation into a new module to better organize the code and tests. I think the Metadata cache could probably also be isolated into its own module, but I didn't want to get too heavy handed in this PR.

Reply:
I this something you are willing to do, or shall I make one?

Reply:
Yes, I'm happy to make an issue for this!

Reply:
Issue created here:
 - https://github.com/apache/datafusion/issues/19074